libs/api/domains/powerbi/src/lib/powerbi.service.ts

```diff
@@ -0,0 +1,89 @@
+import { Configuration, ConfidentialClientApplication } from '@azure/msal-node'
+import { ConfigType } from '@island.is/nest/config'
+import { PowerBiConfig } from './powerbi.config'
+
+type Owner = 'Fiskistofa'
+const BASE_URL = 'https://api.powerbi.com/v1.0/myorg'
+
+export class PowerBiService {
+  constructor(private config: ConfigType<typeof PowerBiConfig>) {}
+
+  async getAccessToken(owner: Owner) {
+    let clientId = ''
+    let clientSecret = ''
+    let tenantId = ''
+
+    switch (owner) {
+      case 'Fiskistofa':
+        {
+          clientId = this.config.fiskistofaPowerBiClientId
+          clientSecret = this.config.fiskistofaPowerBiClientSecret
+          tenantId = this.config.fiskistofaPowerBiTenantId
+        }
+        break
+    }
+
+    const msalConfig: Configuration = {
+      auth: {
+        clientId,
+        clientSecret,
+        authority: `https://login.microsoftonline.com/${tenantId}`,
+      },
+    }
+
+    const clientApplication = new ConfidentialClientApplication(msalConfig)
+
+    const tokenResponse = await clientApplication.acquireTokenByClientCredential(
+      {
+        scopes: ['https://analysis.windows.net/powerbi/api/.default'],
+      },
+    )
+    return tokenResponse?.accessToken as string
+  }
+
+  async getReport(workspaceId: string, reportId: string, accessToken: string) {
+    const url = `${BASE_URL}/groups/${workspaceId}/reports/${reportId}`
+
+    const response = await fetch(url, {
+      method: 'GET',
+      headers: {
+        'Content-Type': 'application/json',
+        Authorization: `Bearer ${accessToken}`,
+      },
+    })
+
+    return response.json()
```

Comment:
This is too little error handling, where non-200 responses will cause hard to trace errors (either `response.json()` fails, or graphql complaints about missing attributes).

I recommend switching to enhanced fetch. It will track request metrics in datadog, automatically log request errors and non-200 responses, and throw fancy FetchError for non-200 responses.

Just add this class member:

```
fetch = createEnhancedFetch({ name: 'PowerBiService' })
```

Then use it here instead:

```suggestion
    const response = await this.fetch(url, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${accessToken}`,
      },
    })

    return response.json()
```

You don't need to change anything else unless you want to `catch` and handle "some" non-200 responses differently.

Reply:
Thanks for the tips, I appreciate it.

I for some reason haven't been able to grasp how exactly it'd be best to handle potential errors. Whether to do nothing at all or try and catch everything.
I guess the answer is that it's good to only try and catch errors if they need to be handled in a specific way or to write out more specific and detailed log messages but then after that it's best to just throw the error instead of letting the code continue doing it's thing by returning null or something instead.

I guess from the client's point of view there either was an error or there wasn't an error so there might be no need to try and return something that indicates what kind of error happened either.

Reply:
Yeah. It can be good to distinguish "expected" errors from "unexpected" errors, as well as client errors (bad input data) vs "internal server errors".

Expected errors should be returned in a way that the client can easily tell them a part and handle them. In REST terms this can be fancy ProblemJSON responses. In GraphQL this may be returning "null" for missing data or having a special Payload schema with typed problem details.

Everything else should be thrown.

The problem I was commenting on in this thread is that when using normal fetch in this way, unexpected errors will be thrown allright, but they won't be very descriptive. Either "string is not json" or "can not read property of undefined". The standard fetch api requires a lot of error handling boilerplate, which we don't want. Thus we use Enhanced Fetch to handle all of the boilerplate so we can keep the code simple like this. 👍

Reply:
Totally :)